crates/sp-lightclient/src/lib.rs

```diff
@@ -79,22 +81,140 @@ pub trait Storage<Header: HeaderT> {
     /// Segment size
     fn segment_size(&self) -> SegmentSize;
 
+    /// K depth
+    fn k_depth(&self) -> NumberOf<Header>;
+
     /// Queries a header at a specific block number or block hash
-    fn header(&self, query: HashOf<Header>) -> Option<HeaderExt<Header>>;
+    fn header(&self, hash: HashOf<Header>) -> Option<HeaderExt<Header>>;
 
     /// Stores the extended header.
     /// as_best_header signifies of the header we are importing is considered best
     fn store_header(&mut self, header_ext: HeaderExt<Header>, as_best_header: bool);
 
-    /// Returns the best known tip of the chain
+    /// Returns the best known tip of the canonical chain
     fn best_header(&self) -> HeaderExt<Header>;
+
+    /// Finalizes the header, prunes and returns all the fork headers at that number
+    fn finalize_header(&mut self, hash: HashOf<Header>) -> Vec<HeaderExt<Header>>;
```

Comment:
> BTW, why is HeaderImporter even a trait? It seems like it should have been a struct instead since it has no methods that need to be implemented.

Fair. The trait was a remnant when it has an associate type for storage.

> Methods like ancestor_of_header_at_depth below are just convenient wrappers, you can either inline them (there is just one place where it is used anyway) or moved to HeaderImporter or something.

This was intentional since I wanted the current logic to assume storage constraint and maybe a specific storage implementation can have some sort of indexing for faster retrieval. Thinking on this now, not a strong point to have it as it.